Microsoft.NET.Build.Containers/net472Definitions.cs

```diff
@@ -0,0 +1,36 @@
+using System;
+using System.IO.Ports;
+using System.Diagnostics.CodeAnalysis;
+using System.Text.RegularExpressions;
+
+namespace System.Runtime.CompilerServices
+{
+    internal static class IsExternalInit { }
+}
+
+namespace Microsoft.NET.Build.Containers
+{
+    /// <summary>
+    /// Specifies that when a method returns <see cref="ReturnValue"/>, the parameter will not be null even if the corresponding type allows it.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Parameter, Inherited = false)]
+    sealed class NotNullWhenAttribute : Attribute
+    {
+        /// <summary>Initializes the attribute with the specified return value condition.</summary>
+        /// <param name="returnValue">
+        /// The return value condition. If the method returns this value, the associated parameter will not be null.
+        /// </param>
+        public NotNullWhenAttribute(bool returnValue) => ReturnValue = returnValue;
+
+        /// <summary>Gets the return value condition.</summary>
+        public bool ReturnValue { get; }
+    }
+
+    public enum PortType
+    {
+        tcp,
+        udp
+    }
+
+    public record Port(int number, PortType type);
```

Comment:
This is to avoid a clash between files. Port is defined in `Image.cs` and importing that file into the net472 build would cause all sorts of other importing/net7.0 issues. I'll create a separate `ContainerTypes` class to house this stuff, and keep the `IsExternalInit` logic (to work around the net472 record type issue)/

Reply:
Update: I just placed it in Containerhelpers to reduce clutter